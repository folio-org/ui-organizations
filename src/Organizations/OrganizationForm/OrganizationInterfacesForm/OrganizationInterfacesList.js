import React from 'react';
import PropTypes from 'prop-types';
import { get } from 'lodash';
import {
  FormattedMessage,
  useIntl,
} from 'react-intl';

import {
  Button,
  Icon,
  MultiColumnList,
  NoValue,
  TextLink,
} from '@folio/stripes/components';
import { Pluggable } from '@folio/stripes/core';
import { acqRowFormatter } from '@folio/stripes-acq-components';

import {
  ADD_INTERFACE_URL,
  EDIT_INTERFACE_URL,
  VIEW_INTERFACE_URL,
} from '../../../interfaces/constants';
import {
  columnMapping,
  visibleColumns,
} from './constants';

const getInterfaceUrl = (orgId, interfaceId) => {
  const ending = interfaceId ? `/interface/${interfaceId}/${VIEW_INTERFACE_URL}` : `/interface/${EDIT_INTERFACE_URL}`;
  const starting = orgId ? `/organizations/${orgId}` : '/organizations';

  return `${starting}${ending}`;
};

const AddInterfaceButton = ({ fetchInterfaces, fields, orgId }) => {
  const addInterfaces = (interfaces = []) => {
    const addedInterfacesIds = new Set(fields.value);
    const newInterface = interfaces.filter(({ id }) => !addedInterfacesIds.has(id));

    if (newInterface.length) {
      const interfaceIds = newInterface.map(inface => inface.id);

      fetchInterfaces([...addedInterfacesIds, ...interfaceIds]);
      interfaceIds.forEach(id => fields.push(id));
    }
  };

  const renderNewInterfaceBtn = () => {
    const url = `/organizations${orgId ? '/' + orgId : ''}/interface/${ADD_INTERFACE_URL}`;

    return (
      <Button
        marginBottom0
        buttonStyle="primary"
        to={url}
      >
        <FormattedMessage id="ui-organizations.interface.newInterface" />
      </Button>
    );
  };

  return (
    <Pluggable
      aria-haspopup="true"
      type="find-interface"
      searchLabel={<FormattedMessage id="ui-organizations.interface.addInterface" />}
      searchButtonStyle="default"
      disableRecordCreation
      addInterfaces={addInterfaces}
      renderNewInterfaceBtn={renderNewInterfaceBtn}
    >
      <span data-test-add-interface>
        <FormattedMessage id="ui-organizations.interface.noFindInterfacePlugin" />
      </span>
    </Pluggable>
  );
};

AddInterfaceButton.propTypes = {
  fetchInterfaces: PropTypes.func.isRequired,
  fields: PropTypes.object,
  orgId: PropTypes.string,
};

const alignRowProps = { alignLastColToEnd: true };

const OrganizationInterfacesList = ({ fetchInterfaces, fields, interfaces, orgId }) => {
  const intl = useIntl();
  const contentData = fields.value.map((interfaceId, _index) => ({
    ...get(interfaces, interfaceId, {}),
    _index,
  }));

  const anchoredRowFormatter = ({ rowProps, ...rest }) => {
    return acqRowFormatter({
      ...rest,
      rowProps: {
        ...rowProps,
        to: getInterfaceUrl(orgId, rest.rowData.id),
      },
    });
  };

  const resultsFormatter = {
    interfaceName: ({ name }) => name,
    interfaceUrl: (item) => (item.uri
      ? (
        <TextLink
          rel="noopener noreferrer"
          target="_blank"
          href={item.uri}
        >
          {item.uri}
        </TextLink>
      )
      : <NoValue />
    ),
    unassignInterface: (item) => (
      <Button
        align="end"
        aria-label={intl.formatMessage({ id: 'ui-organizations.interface.button.unassign' })}
        buttonStyle="fieldControl"
        data-test-unassign-interface
        type="button"
        onClick={(e) => {
          e.preventDefault();
          fields.remove(item._index);
        }}
      >
        <Icon icon="times-circle" />
      </Button>
    ),
  };

  return (
    <>
      <MultiColumnList
        id="interface-list"
        columnMapping={columnMapping}
        contentData={contentData}
        formatter={resultsFormatter}
        rowFormatter={anchoredRowFormatter}
        rowProps={alignRowProps}
        visibleColumns={visibleColumns}
      />
      <br />
      <AddInterfaceButton
        fetchInterfaces={fetchInterfaces}
        fields={fields}
        orgId={orgId}
      />
    </>
  );
};

OrganizationInterfacesList.propTypes = {
  fetchInterfaces: PropTypes.func.isRequired,
  fields: PropTypes.object,
  orgId: PropTypes.string,
  interfaces: PropTypes.object,
};

export default OrganizationInterfacesList;
