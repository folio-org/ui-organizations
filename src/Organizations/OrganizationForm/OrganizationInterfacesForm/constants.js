export const columnMapping = {
  name: 'Name',
  url: 'URL',
  unassign: null,
};

export const visibleColumns = ['name', 'url', 'unassign'];
