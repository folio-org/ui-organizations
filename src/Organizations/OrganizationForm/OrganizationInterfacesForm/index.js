export { default as OrganizationInterfacesForm } from './OrganizationInterfacesForm';
