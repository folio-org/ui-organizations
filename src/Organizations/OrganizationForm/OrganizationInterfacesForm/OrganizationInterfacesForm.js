import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { FieldArray } from 'react-final-form-arrays';

import {
  Col,
  Loading,
  Row,
} from '@folio/stripes/components';
import { stripesConnect } from '@folio/stripes/core';
import { batchFetch } from '@folio/stripes-acq-components';

import { interfacesResource } from '../../../common/resources';
import OrganizationInterfacesList from './OrganizationInterfacesList';

function OrganizationInterfacesForm({ open, orgId, mutator, storedInterfaces }) {
  const [interfaces, setInterfaces] = useState();
  const [isLoading, setIsLoading] = useState(true);

  const refreshInterfaces = useCallback(interfaceIds => {
    setIsLoading(true);
    batchFetch(mutator.interfacesManualFetch, interfaceIds || [])
      .then(setInterfaces)
      .catch(() => {
        setInterfaces([]);
      })
      .finally(() => setIsLoading(false));
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => refreshInterfaces(storedInterfaces), [refreshInterfaces, storedInterfaces]);

  if (isLoading || !open) {
    return (
      <Loading />
    );
  }

  const interfacesMap = interfaces.reduce((acc, item) => {
    acc[item.id] = item;

    return acc;
  }, {});

  return (
    <Row>
      <Col xs={12}>
        <FieldArray
          name="interfaces"
          component={OrganizationInterfacesList}
          interfaces={interfacesMap}
          fetchInterfaces={refreshInterfaces}
          orgId={orgId}
        />
      </Col>
    </Row>
  );
}

OrganizationInterfacesForm.manifest = Object.freeze({
  interfacesManualFetch: {
    ...interfacesResource,
    fetch: false,
    accumulate: true,
  },
});

OrganizationInterfacesForm.propTypes = {
  mutator: PropTypes.object.isRequired,
  open: PropTypes.bool,
  orgId: PropTypes.string,
  storedInterfaces: PropTypes.arrayOf(PropTypes.string),
};

export default stripesConnect(OrganizationInterfacesForm);
