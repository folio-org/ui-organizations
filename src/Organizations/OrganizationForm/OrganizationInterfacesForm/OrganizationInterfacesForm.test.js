import React from 'react';
import { render, screen } from '@folio/jest-config-stripes/testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import stripesFinalForm from '@folio/stripes/final-form';

import { orgInterface } from 'fixtures';

import OrganizationInterfacesForm from './OrganizationInterfacesForm';

const mutatorMock = {
  interfacesManualFetch: {
    GET: jest.fn().mockReturnValue(Promise.resolve([orgInterface])),
  },
};
const TestForm = stripesFinalForm({})(
  () => {
    return (
      <form>
        <OrganizationInterfacesForm
          open
          stripes={{}}
          storedInterfaces={[orgInterface.id]}
          mutator={mutatorMock}
        />
      </form>
    );
  },
);

const defaultProps = {
  initialValues: {
    interfaces: [orgInterface.id],
  },
};
const renderForm = (props = defaultProps) => render(
  <TestForm
    onSubmit={jest.fn()}
    {...props}
  />,
  { wrapper: MemoryRouter },
);

describe('OrganizationInterfacesForm', () => {
  beforeEach(() => {
    global.document.createRange = global.document.originalCreateRange;
  });

  afterEach(() => {
    global.document.createRange = global.document.mockCreateRange;
  });

  it('should render correct structure', async () => {
    const { asFragment } = renderForm();

    await screen.findByText('ui-organizations.interface.name');

    expect(asFragment()).toMatchSnapshot();
  });
});
