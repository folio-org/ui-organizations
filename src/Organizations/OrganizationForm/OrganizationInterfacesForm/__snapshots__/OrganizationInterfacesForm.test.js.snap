// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationInterfacesForm should render correct structure 1`] = `
<DocumentFragment>
  <form>
    <div
      class="row"
    >
      <div
        class="col-xs-12"
      >
        <div
          aria-live="assertive"
          aria-relevant="additions"
          class="sr-only"
          tabindex="-1"
        />
        <div
          aria-rowcount="2"
          class="mclContainer"
          data-total-count="0"
          id="interface-list"
          role="grid"
          style="position: relative; overflow: hidden; height: auto; width: 100%;"
          tabindex="0"
        >
          <div
            class="mclHeaderContainer"
          >
            <div
              aria-rowindex="1"
              class="mclHeaderRow"
              role="row"
              style="display: flex;"
            >
              <div
                aria-sort="none"
                class="mclHeader"
                id="list-column-interfacename"
                role="columnheader"
                style="flex: 0 0 auto;"
              >
                ui-organizations.interface.name
              </div>
              <div
                aria-sort="none"
                class="mclHeader"
                id="list-column-interfaceurl"
                role="columnheader"
                style="flex: 0 0 auto;"
              >
                ui-organizations.interface.url
              </div>
              <div
                aria-sort="none"
                class="mclHeader"
                id="list-column-unassigninterface"
                role="columnheader"
                style="flex: 0 0 auto;"
              />
            </div>
          </div>
          <div
            class="mclScrollable"
            style="position: static; width: 100%;"
          >
            <div
              class="mclRowContainer"
              role="rowgroup"
              style="position: static;"
            >
              <div
                aria-rowindex="2"
                class="mclRowFormatterContainer"
                data-row-index="row-0"
                role="row"
                style="position: static;"
              >
                <a
                  aria-label=""
                  class="mclRow mclIsOdd mclIsInteractive alignLastColToEnd"
                  data-row-inner="0"
                  href="/organizations/interface/793b9d42-ae12-41d7-a36a-f33e5b9e82c5/view"
                  style="min-width: 0px;"
                  tabindex="0"
                >
                  <div
                    class="mclCell  mclCellStyle"
                    role="gridcell"
                  >
                    GOBI
                  </div>
                  <div
                    class="mclCell  mclCellStyle"
                    role="gridcell"
                  >
                    www.gobi3.com
                  </div>
                  <div
                    class="mclCell  mclCellStyle"
                    role="gridcell"
                  >
                    <button
                      aria-label="ui-organizations.interface.button.unassign"
                      class="button fieldControl align-end"
                      data-test-unassign-interface="true"
                      type="button"
                    >
                      <span
                        class="inner"
                      >
                        <span>
                          Icon
                        </span>
                      </span>
                    </button>
                  </div>
                </a>
              </div>
              <div
                class="mclEndOfListContainer"
                data-end-of-list="1"
              >
                <div
                  class="mclCenteredContainer"
                  style="width: 100%; visibility: hidden; height: 0px; padding: 0px;"
                >
                  <div
                    class="textCentered"
                    data-test-layout="true"
                  >
                    <div
                      class="mclEndOfList"
                      data-test-layout="true"
                    >
                      <span>
                        Icon
                      </span>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <br />
        <span
          data-test-add-interface="true"
        >
          ui-organizations.interface.noFindInterfacePlugin
        </span>
      </div>
    </div>
  </form>
  <div />
</DocumentFragment>
`;
